rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,204 @@
+dnnl_memory_desc_t scalar_md;
+dnnl_memory_desc_create_host_scalar(
+        &scalar_md, dnnl_f32);
+
+// Create a memory object for the scalar
+dnnl_memory_t scalar_mem;
+dnnl_status_t status = dnnl_memory_create_host_scalar(&scalar_mem, scalar_md, &value);
+
+// Use as regular memory object in execute function...
+```
+
+```C
+// C++ API
+
+// Scalar value
+float value = 42.0f;
+
+// Create a memory object for the scalar (no engine needed)
+memory scalar_mem(memory::desc::host_scalar(memory::data_type::f32), &value);
+
+// Use as regular memory object in execute function...
+```
+
+POC for API changes is available in the draft PR [^3],
+you could check it out for simple example, API tests, and some implementation
+details (although the implementation is not complete yet).
+
+Pros:
+- Users do not need to recreate primitives when scalar values change.
+- Allows passing scalars as kernel arguments, minimizing performance overhead.
+
+Cons:
+- Requires explicit support in each non-host implementation to handle host-side scalars
+    correctly.
+
+## Other options considered
+
+### Option 1: User to rely on USM `malloc_host` (no changes to oneDNN)
+
+With this option, the user allocates a USM memory buffer on the host using
+`malloc_host`, stores the scalar value in it, and creates a `dnnl::memory`
+object from that buffer, which can then be passed to the execute function.
+
+Pros:
+- Allows to pass host-side scalars without changing the API or internals of
+    oneDNN.
+
+Cons:
+- Requires user to keep the USM memory alive until kernel computations finish.
+- There is a latency overhead when the GPU fetches host USM data. While the data
+    may be cached in GPU memory after the initial read, that first access can
+    introduce a delay.
+- **Update:** After testing this approach, it was found that the overhead is still
+    noticeable and managing the lifetime of USM memory can be complex for applications.
+
+### Option 2: Take host-side scalars as primitive descriptor parameters
+
+Here user would pass the host side scalar as a parameter to primitive creation,
+as is done today. The main caveats here are:
+- It requires user to recreate primitive descriptors and primitives every time
+    the host-side scalar changes value.
+- If the implementation jits a kernel that depends on that parameter, we can get
+    large overheads due to re-jitting.
+
+Pros:
+- oneDNN API can already handle this for parameters that do not frequently
+    change.
+
+Cons:
+- Forces user to recreate primitive descriptor and primitive objects every-time
+    that host-side scalar value changes.
+- Potential extra jitting overhead if internal implementation considers that
+    parameter constant.
+
+This option is not recommended if the host-side scalar value is expected to
+frequently change.
+
+### Option 3: Take engine kind in memory descriptor
+
+Here, we would allow users to pass any memory on host or device. This would
```

Comment:
I'm a bit puzzled by this option. Suppose memory descriptors become engine-aware, what impact would that have? The primitive descriptor would still be created for a particular engine and memory objects would still be created for a particular engine. If we allow passing memory objects created for different engines then how would making memory descriptors engine-aware help? Would it just be to validate that the engine kinds in memory descriptors and memory objects are consistent?

In any case, I believe that making memory descriptors engine-aware is a deal-breaker as they were designed to be engine-agnostic and contain no engine-specific information. This, in my opinion, is a major drawback of this option so I just want to make sure I understand this option correctly.